Fix some Scrutinizer issues and improve code quality

pbricout opened the pull request:
Based on the following report:
https://scrutinizer-ci.com/g/Zazama/node-id3/code-structure/master

See commit by commit for more details.

Review with **Hide Whitespace changes** checked.

I did not rework the following functions in `index.js` yet:

- update
- getTagsFromBuffer
- getFramesFromID3Body
- getTagsFromFrames

but there is already quite a lot changes in this PR and it fixes 23 issues reported by Scrutinizer, so I am moving it to ready for review.

I have probably started to diverge from the original coding style and introduce some of my coding style, starting to introduce more `const`, split large blocks, remove redundancy, use more functional still (total size with reduce for example), etc.

I started to add more types and documentations, but there is still a lot to do of course, the goal was not to address everything but to improve some parts.

Let me know if you are happy with these changes.

Thanks

Zazama commented on index.js:
```diff
@@ -419,7 +419,8 @@ module.exports.removeTagsFromBuffer = function(data) {
 
-    if ((hSize[0] | hSize[1] | hSize[2] | hSize[3]) & 0x80) {
+    const isMsbSet = !!((hSize[0] | hSize[1] | hSize[2] | hSize[3]) & 0x80)
+    if (!isMsbSet) {
```
I think the tests fail because it should be `isMsbSet` without the `!`

pbricout replied:
Ah yes, you are right, I have been confused by the comment below as the comment was having a **not** in the wording! This should be fixed now.

Zazama commented:
I think it is interesting that Scrutinizer drops the quality rating between master and your changes.

It seems to me that even if you refactor the functions into smaller parts, Scrutinizer still analyzes the full path that the function takes, including the calls in it (which, I guess, is why readSync is being rated with F): https://scrutinizer-ci.com/g/Zazama/node-id3/inspections/2dc61913-e67c-45f9-baa5-06b237a39b11/code-structure/js-function/index.js%7CreadSync
I can't really explain why it would do so otherwise.

It feels like the rating is kind of random, e.g. this commit made the rating go up by 0.23 points (from 3.69 to 3.92): https://github.com/Zazama/node-id3/commit/2b6a7fac800d3b80c8ebbf6a613244520d90528e
and I don't see why the rating shouldn't improve by your changes.

Just an observation, not really important 🙂

pbricout commented:
> I think it is interesting that Scrutinizer drops the quality rating between master and your changes.
Yes, I noticed that too, I really don't understand why. I did not investigate much, one reason I was thinking was that the file was getting larger, but otherwise I am not sure.

> It seems to me that even if you refactor the functions into smaller parts, Scrutinizer still analyzes the full path that the function takes, including the calls in it (which, I guess, is why readSync is being rated with F): https://scrutinizer-ci.com/g/Zazama/node-id3/inspections/2dc61913-e67c-45f9-baa5-06b237a39b11/code-structure/js-function/index.js%7CreadSync I can't really explain why it would do so otherwise.
> 
> // Edit: that doesn't really make sense because readAsync get's A-rated: https://scrutinizer-ci.com/g/Zazama/node-id3/inspections/2dc61913-e67c-45f9-baa5-06b237a39b11/code-structure/js-function/index.js%7CreadAsync

Looking at both, it seems that the complexity (i.e. conditions/cyclomatic complexity) rating is buggy.
`readAsync` has two nested if/else and **conditions** is reported as 4 which seems correct, however `readSync` has only one if/else but conditions is reported as 28 which seems completely wrong.

> // Edit2: Codeclimate, for example. will just mark read's complexity as fixed and not complain about the new functions: https://codeclimate.com/github/Zazama/node-id3/compare/scrut-test

I did not know about this site, it seems a bit more opinionated about what the criteria of code quality (number of lines per function/module for example). I did explore more yet.

pbricout commented:
If you are OK with these changes, I will push a bit more changes.

pbricout commented on src/ID3Util.js:
```diff
@@ -78,15 +78,18 @@ module.exports.getSpecOptions = function(specName, version) {
     }
-    return true;
-};
+    if((buffer[6] | buffer[7] | buffer[8] | buffer[9]) & 128 === 1) {
```
I really don't understand this test, surely (128 === 1) can never be true?
Removing it, still pass tests.
I was thinking the original intent was to test the size is valid, however replacing this test with `return this.isValidEncodedSize(buffer.slice(6, 10))` make the tests to fail.

pbricout commented:
Ready for review.

pbricout commented on index.js:
```diff
@@ -164,24 +179,10 @@ module.exports.read = function(filebuffer, options, fn) {
-            fn(null, this.getTagsFromBuffer(filebuffer, options))
-        }
+    if(isFunction(fn)) {
+        return readAsync.bind(this)(filebuffer, options, fn)
```
I am not super keen by this `bind(this)`, this is a **hack** as `readAsync` is not part of the same namespace that it is bound to since `readAsnync` and `readSync` are not exported.

Addressing this properly, would require to not export directly all the functions (same I did for `removeTagsFromBuffer`). It makes it a bit harder to see what is exported when scattered all other the file.
In that case  a single export at the end with all the exported would probably be nicer.

Zazama replied:
There is no state in the NodeID3 "class", so it would make sense to just define all functions without module.exports and then export the public ones at the end so it's not mixed up.

pbricout replied:
OK, I'll do that later this evening.

pbricout replied:
I pushed a massive commit with the changes, it required quite a lot of changes, but this is way better now.
Tests are passing, I had to fix one test.

pbricout commented on src/ID3Frames.js:
```diff
@@ -395,9 +399,10 @@ module.exports.WXXX = {
-        data.keyEvents.forEach((keyEvent) => {
+        const builder = new ID3FrameBuilder("ETCO")
+            .appendStaticNumber(data.timeStampFormat, 1)
+        const keyEvents = data.keyEvents || []
```
I am not sure why you wanted to make `keyEvents` optional, I don't see the benefit, we surely want to write events. I would not make it optional.

pbricout commented on index.js:
```diff
@@ -503,28 +503,37 @@ module.exports.removeTags = function(filepath, fn) {
+    return (a, b) => fn(b, a)
+}
+
+// The reorderParameter is a workaround because the callback function
+// does not have a consistent interface between all the API functions.
+// Ideally, all the functions should align with the promise style and
+// always have the result first and the error second.
+// Changing this would break the current public API.
+// This could be changed internally and swap the parameter in a light
+// wrapper when creating the public interface and then remove in a
+// version 1.0 later with an API breaking change.
```
This big comment is cluttering a bit the code and could be in the commit message instead, let me know what you think.

Zazama replied:
I'd leave it in for now because there are multiple comments in the code about v1 changes, we could extract everything at once into an issue or something like that later.

pbricout replied:
OK, thanks.

pbricout commented:
It seems that a lot of functions that are actually currently exported in `index.js` are not supposed to be public.
I will move them to a new module so they can be shared in `ID3Frames.js`

Zazama commented:
I will test everything carefully tomorrow and merge the PR. I think we should stop changing more in this PR so there won't be more conflicts and create new ones if there's something else to do.

pbricout commented:
It's interesting that you started to add support for v4, this is exactly what I had in mind to do next, because I needed support for recording (TDRC) and release time (TDRL), thanks for starting to work on it.

pbricout commented:
> I will test everything carefully tomorrow and merge the PR. I think we should stop changing more in this PR so there won't be more conflicts and create new ones if there's something else to do.

I agree, I should stop adding changes for this PR.
I have rebased, fixed the conflicts and forced pushed to have a clean PR.
There a few other things that needs to be reviewed, but I definitely address them in a new PR.
Thanks for patience, reviewing, testing and accepting the changes.

pbricout commented on index.js:
```diff
@@ -246,313 +214,122 @@ module.exports.update = function(tags, filebuffer, options, fn) {
-        })
-    }
+const PromiseExport = {
+    create: (tags) => makePromise(create.bind(null, tags), makeSwapParameters),
+    write: (tags, file) => makePromise(write.bind(null, tags, file)),
+    update: (tags, file) => makePromise(update.bind(null, tags, file)),
+    read: (file, options) => makePromise(read.bind(null, file, options)),
+    removeTags: (filepath) => makePromise(removeTags.bind(null, filepath))
```
Interestingly Scrutinizer has a false positive here and tag these `bind`s as unnecessary because there is no `this` parameters, but completely miss the points that more parameters are bound!

Zazama replied:
I've reported it as a bug and got as reply: `Hi Jan, thanks for the report. We already had an internal issue for this, I've added your report to it and will keep you updated about its progress. If you need anything else, just reply to this ticket. `

If they don't fix it soon, I can just hide this issue.

pbricout replied:
Excellent to have reported it, thanks.

pbricout commented on index.js:
```diff
@@ -1,60 +1,98 @@
+    return data
+}
+
+function writeInBuffer(tags, buffer, fn) {
+    buffer = removeTagsFromBuffer(buffer) || buffer
+    const completeBuffer = Buffer.concat([tags, buffer])
+    if(isFunction(fn)) {
+        fn(null, completeBuffer)
+        return undefined
+    }
+    return completeBuffer
+}
```
This is done differently than `read` where the buffer case is handled inside `readAync` and `readSync`, in another PR we should probably align it with `read`.

Zazama replied:
I've changed it here because I got confused when reviewing, should be the same as read now

Zazama commented:
> It's interesting that you started to add support for v4, this is exactly what I had in mind to do next, because I needed support for recording (TDRC) and release time (TDRL), thanks for starting to work on it.

I've added the text frames for now because I already did most of the work some months ago but never finalized it.

> Thanks for patience, reviewing, testing and accepting the changes.

Thank you for the PRs, they were very good 🙂 
And the rating went from 3.57 to 4.82 (https://scrutinizer-ci.com/g/Zazama/node-id3/statistics/)

I think we can push a new version now, or is something missing?

pbricout commented on src/ID3Helpers.js:
```diff
@@ -197,7 +197,7 @@ function getTagsFromFrames(frames, ID3Version, options = {}) {
         let decoded
         if(ID3Frames[specName]) {
-            decoded = ID3Frames[specName].read(frame.body, ID3Version, module.exports)
+            decoded = ID3Frames[specName].read(frame.body, ID3Version)
```
Good spot, I forgot to remove that one.

pbricout commented:
> And the rating went from 3.57 to 4.82 (https://scrutinizer-ci.com/g/Zazama/node-id3/statistics/)

Strange I don't see `ID3Helpers.js` in the report ?
https://scrutinizer-ci.com/g/Zazama/node-id3/code-structure/master

> I think we can push a new version now, or is something missing?

I pushed  a quick small PR to fix one more issue:
https://github.com/Zazama/node-id3/pull/129

I think we should also integrate this contribution PR, I can download it fix the conflicts and create a new one:
https://github.com/Zazama/node-id3/pull/116

Zazama commented:
> I think we should also integrate this contribution PR, I can download it fix the conflicts and create a new one:

That would be good, but I think it doesn't make sense to use the PR as it is because it changes file spacing etc., so it might be better to just copy the relevant bits into a new one.

It is also not 100% spec because the identifier is binary data and shouldn't assumed to be a string in my opinion.

```javascript
identifier: reader.consumeStaticValue('string')

// should, like in the PRIV frame, be

identifier: reader.consumeStaticValue()
```

Also interesting:
>  The 'Owner identifier' is then followed by the actual identifier, which may be up to 64 bytes.

I'm not sure if the library should handle cutting the binary data off at 64 bytes or if that should be the user's responsibility.

pbricout commented:
> That would be good, but I think it doesn't make sense to use the PR as it is because it changes file spacing etc., so it might be better to just copy the relevant bits into a new one.

That was my plan, that's why I said I would re-create a PR.

> It is also not 100% spec because the identifier is binary data and shouldn't assumed to be a string in my opinion.

I will check, thanks

> I'm not sure if the library should handle cutting the binary data off at 64 bytes or if that should be the user's responsibility.

That's a good point, from my understanding, most of the time the library try to be quite thin, if we truncate ourselves, the user might not know it is truncated, there are currently plenty of other constraints that the library does not check, I think at this point we should just document it.
